### PR TITLE
KothaEtAl2020regional

### PR DESCRIPTION
**Pull request by Shreyasvi91**
A new gsim class has been proposed wherein the Site-specific (delta_c3 and its standard error) and source-specific (delta_l2l and its standard error) values are automatically selected and applied in the calculation.

### CODE REVIEW

**Review by mmpagani — CHANGES_REQUESTED** [review]

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 1]
```diff
@@ -22,16 +22,23 @@
...
 from scipy.constants import g
-
+from openquake.hazardlib.geo.packager import fiona
 from openquake.baselib.general import CallableDict
 from openquake.hazardlib.gsim.base import GMPE, CoeffsTable
 from openquake.hazardlib import const
 from openquake.hazardlib.imt import PGA, PGV, SA, from_string
 from openquake.hazardlib.gsim.nga_east import (get_tau_at_quantile, ITPL,
                                                TAU_EXECUTION, TAU_SETUP)
+from shapely.geometry import shape, Point, Polygon
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:37]: Okay. Shapely imports have been moved above and the order is Point, Polygon, and shape.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 2]
```diff
@@ -22,16 +22,23 @@
...
+import itertools
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:38]: itertools has been removed.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 3]
```diff
@@ -161,6 +191,29 @@ def get_magnitude_scaling(C, mag):
...
+    if delta_l2l_epsilon is provided, standard error of delta_c3 
+    will be included. In this case epsilon represents the quantile
+    distribution. Hence, a epsilon value of +/- 1.6 represents
+    the 5th and 95th percentile of the dl2l values. 
+    """
+
+    dl2l = np.array([[0] * 2] * len(ctx.hypo_lon), dtype = float)
+    fname = os.path.join(DATA_FOLDER, 'kotha_tectonic_regions.geojson')
+    with fiona.open(fname) as tec:
+        for i, (lo, la) in enumerate(zip(ctx.hypo_lon, ctx.hypo_lat)):
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:207]: Thanks for the suggestion. I will look into this.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 4]
```diff
@@ -91,29 +98,52 @@ def _get_h(C, hypo_depth):
...
+    In this case epsilon represents the quantile distribution. 
+    Hence, a epsilon value of +/- 1.6 represents the 5th and 95th 
+    percentile of the C3 values. 
+    """
+    delta_c3 = np.array([[0] * 2] * len(sctx.lon), dtype = float)
+    fname = os.path.join(DATA_FOLDER, 'kotha_attenuation_regions.geojson')
+    with fiona.open(fname) as att:
+        for i, (lo, la) in enumerate(zip(sctx.lon, sctx.lat)):
+            pt = Point(lo, la)
+            region = []
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:133]: removed.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 5]
```diff
@@ -91,29 +98,52 @@ def _get_h(C, hypo_depth):
...
 def get_distance_coefficients_1(kind, c3, c3_epsilon, C, imt, sctx):
     """
     Returns either the directly specified c3 value or the c3 from the
     existing tau_c3 distribution
     """
     if c3:
+        
         # Use the c3 that has been defined on input
-        return c3
+        return c3[imt][0]
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:110]: changed back

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 6]
```diff
@@ -92,13 +98,14 @@ def _get_h(C, hypo_depth):
 get_distance_coefficients = CallableDict()
 
 
-@get_distance_coefficients.add("base", "site", "slope")
+@get_distance_coefficients.add("base","site","slope")
```
Comment (mmpagani) [inline openquake/hazardlib/gsim/kotha_2020.py:101]: You did not change it as I suggested.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 7]
```diff
@@ -139,12 +145,37 @@ def get_distance_coefficients_2(kind, c3, c3_epsilon, C, imt, sctx):
     return c3_ + c3_epsilon * tau_c3
 
 
-def get_distance_term(kind, c3, c3_epsilon, C, ctx, rjb, imt):
+@get_distance_coefficients.add("regional")
+def get_distance_coefficients_3(kind, c3, delta_c3_epsilon, C, imt, sctx):
+    """
+    Return site-specific coefficient 'C3'. The function retirieves the
```
Comment (mmpagani) [inline openquake/hazardlib/gsim/kotha_2020.py:151]: retirieves -> retrieves

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 8]
```diff
@@ -139,12 +145,37 @@ def get_distance_coefficients_2(kind, c3, c3_epsilon, C, imt, sctx):
...
+def get_distance_coefficients_3(kind, c3, delta_c3_epsilon, C, imt, sctx):
+    """
+    Return site-specific coefficient 'C3'. The function retirieves the 
+    value of delta_c3 and the standard error of delta_c3 from the 'att' 
+    geojson file depending on the location of site. This delta_c3 is 
+    added to the generic coefficient 'c3' from the GMPE. A delta_c3_epsilon
+    value of +/- 1.6 gives the 95% confidence interval for delta_c3.
+    """
+    s = []
+    [s.append(Point(lon, lat)) for lon, lat in zip(sctx.lon, sctx.lat)]
```
Comment (mmpagani) [inline openquake/hazardlib/gsim/kotha_2020.py:158]: The part of reading the input file should be moved to the __init__ function. In this way, you avoid reading the same file thousand of times.

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 9]
```diff
@@ -162,6 +193,30 @@ def get_magnitude_scaling(C, mag):
...
+    Returns rupture source specific delta_l2l values. The method
+    retrieves the delta_l2l and standard error of delta_l2l values.
+    if delta_l2l_epsilon is provided, standard error of delta_c3 
+    will be included. A delta_l2l_epsilon value of +/- 1.6 gives 
+    the 95% confidence interval for delta_l2l.
+    """
+    f = []
+    [f.append(Point(lon, lat)) for lon, lat in zip(ctx.hypo_lon, ctx.hypo_lat)]
+    dl2l = np.zeros((len(ctx.hypo_lon), 2), dtype = float)
+    fname = os.path.join(DATA_FOLDER, 'kotha_tectonic_regions.geojson')
```
Comment (mmpagani) [inline openquake/hazardlib/gsim/kotha_2020.py:207]: Same suggestion provided for the path term

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 8 above)
Comment (mmpagani) [inline openquake/hazardlib/gsim/kotha_2020.py:158]:
```
s = []
[s.append(Point(lon, lat)) for lon, lat in zip(sctx.lon, sctx.lat)]
```
should be 
```
s = [(Point(lon, lat)) for lon, lat in zip(sctx.lon, sctx.lat)]
```

File: openquake/hazardlib/tests/gsim/kotha_2020_test.py [hunk 10]
```diff
@@ -125,6 +128,41 @@ def test_instantiation_c3_non_dict(self):
...
+        imt = 'PGA'
+        hypo = [[-4.13, 38.55], [7.74, 46.23], [10.579, 62.477], [12.34, 45.03], [15.02, 39.80]]
+        ctx = pd.DataFrame(hypo, columns=['hypo_lon', 'hypo_lat'])
+        expected_val = np.array([0., -0.1490727,  0., -0.28239376, -0.2107627 ])
+        dl2l = get_dl2l(ctx, imt, delta_l2l_epsilon)
+        np.testing.assert_array_equal(dl2l, expected_val)
+    
+
+class KothaEtAl2020regionalTestCase(BaseGSIMTestCase):
+    ##Testing the adjusted version of the GMPE
```
[the remark this answers is not in the document]
Comment (mmpagani) [reply openquake/hazardlib/tests/gsim/kotha_2020_test.py:164]: ?

**Review by mmpagani — CHANGES_REQUESTED** [review]

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 7 above)
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:151]: changed. Thanks.

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 6 above)
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:101]: it has been changed now.

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 8 above)
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:158]: The geojson files have been instantiated in the **init** function now.

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 9 above)
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:207]: incorporated.

File: openquake/hazardlib/gsim/kotha_2020.py (hunk 8 above)
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:158]: incorporated.

File: openquake/hazardlib/tests/gsim/kotha_2020_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/tests/gsim/kotha_2020_test.py:164]: explanation has been added.

**Review by mmpagani — APPROVED** [review]

File: openquake/hazardlib/gsim/kotha_2020.py [hunk 11]
```diff
@@ -368,6 +405,7 @@ def __init__(self, sigma_mu_epsilon=0.0, c3_epsilon=0.0, ergodic=True,
         self.sigma_mu_epsilon = sigma_mu_epsilon
         self.c3_epsilon = c3_epsilon
         self.ergodic = ergodic
+        self.att = None
```
[the remark this answers is not in the document]
Comment (Shreyasvi91) [reply openquake/hazardlib/gsim/kotha_2020.py:408]: actually no it's not necessary. I have removed it.

**Comment by micheles** [comment]
Notice that fiona is NOT a dependency of the engine. @vot4anto has to do some work in order to add fiona to the list of dependencies across all supported platforms and python versions.

**Comment by micheles** [comment]
I understand that there is a deadline, so let's merge this once the fiona dependency is added, but I must point out that the performance will be terrible, since you are doing geospatial queries for each rupture. The right way to implement this feature would be to pre-generate all the coefficients *before* starting the calculation and then retrieve the right coefficients for the given rupture. However this is much more complicated and I cannot do it right now due to other commitments.

**Review by micheles — APPROVED** [review]

**Pull request merged**